src/overmap.cpp

```diff
@@ -3653,7 +3653,12 @@ bool overmap::build_lab( int x, int y, int z, int s, std::vector<point> *lab_tra
         }
     }
     if( generate_stairs && !generated_lab.empty() ) {
-        const point p = random_entry( generated_lab );
+        point p;
+        int tries = 0;
+        do {
+            p = random_entry( generated_lab );
+            tries++;
+        } while( ter( p.x, p.y, z + 1 ) != labt && tries < 15 );
```

Comment:
This method only decreases the likelihood of wrongly placed stairs but there will still be a small chance that the 15 tries all fail and the stairs become wrongly placed. If my understanding on the code is correct, It would be better to generate a list containing those `point`s in `generated_lab` that satisfy `ter( p.x, p.y, z + 1 ) == labt`, and call `random_entry()` on the generated list. This way it is guaranteed to place the stairs at the correct place.